Support CSV column headers written unquoted

craig410 opened the pull request.

craig410 committed:
Support CSV column headers written unquoted

acoulton commented on src/CSV/CSVWriter.php:
```diff
@@ -35,6 +36,10 @@ class CSVWriter
         'write_utf8_bom' => FALSE
     ];
 
+    public function __construct(private readonly bool $quote_headers = true)
```
I think this should be part of the options to `open` for consistency with the `write_utf8_bom` option - as you'd want to specify when writing the file rather than requiring a separate DI service with a different configuration?

acoulton commented on src/CSV/CSVWriter.php:
```diff
@@ -72,7 +77,18 @@ public function write(array $row)
+                array_walk(
+                    $this->expect_schema,
+                    fn(string $str) => ! str_contains($str, ',')
+                        ?: throw new RunTimeException(
```
`RuntimeException`

acoulton commented on src/CSV/CSVWriter.php:
```diff
@@ -72,7 +77,18 @@ public function write(array $row)
             }
             $this->expect_schema = $row_schema;
-            \fputcsv($this->resource, $this->expect_schema);
+            if ($this->quote_headers) {
+                \fputcsv($this->resource, $this->expect_schema);
+            } else {
+                array_walk(
+                    $this->expect_schema,
+                    fn(string $str) => ! str_contains($str, ',')
+                        ?: throw new RunTimeException(
+                            "Column header `$str` cannot contain comma if headers are not quoted"
+                        )
+                );
+                fwrite($this->resource, implode(',', $this->expect_schema).PHP_EOL);
+            }
```
Might be neater to extract a `private function writeHeader($keys)`?

acoulton commented on test/unit/CSV/CSVWriterTest.php:
```diff
@@ -134,6 +136,55 @@ public function test_it_does_not_write_column_headers_before_subsequent_rows()
+    {
+        $row = ['our data' => 'is at times', 'really big' => 'often'];
+
+        return [
```
Probably want a testcase with the default (e.g. `->open($file, [])`) as well as the explicit `['quote_headers' => true]` and `['quote_headers' => false]` cases?

acoulton reviewed (changes requested):
Implementation looks good, just a couple thoughts

acoulton commented on src/CSV/CSVWriter.php:
```diff
@@ -28,11 +29,9 @@ class CSVWriter
-     */
-    protected $options = [
-        'write_utf8_bom' => FALSE
+    protected array $options = [
```
Technically, can't add a strict typehint to a `protected` property without a BC break - if someone is extending the class (e.g. with their own default / custom options) they would have to add the typehint too.

acoulton reviewed (approved):
LGTM if you can revert to a phpdoc typehint

craig410 committed:
Don't run actions build on a schedule

Only causes it to become disabled when there aren't regular pushes and then the tests don't run until you manually enable it.

craig410 committed:
Upgrade github actions step versions
